utils/analysis/factory.go

```diff
@@ -217,3 +227,42 @@ func ValidateMetric(metric v1alpha1.Metric) error {
 	}
 	return nil
 }
+
+func extractValueFromRollout(r *v1alpha1.Rollout, path string) (string, error) {
+	j, err := json.Marshal(r)
+	if err != nil {
+		return "", err
+	}
+
+	m := interface{}(nil)
+	json.Unmarshal(j, &m)
+	sections := regexp.MustCompile("[\\.\\[\\]]+").Split(path, -1)
+	for _, section := range sections {
+		if asArray, ok := m.([]interface{}); ok {
+			if i, err := strconv.Atoi(section); err != nil {
+				return "", fmt.Errorf("invalid index %s", section)
+			} else if i >= len(asArray) {
+				return "", fmt.Errorf("index %d out of range", i)
+			} else {
+				m = asArray[i]
+			}
+		} else if asMap, ok := m.(map[string]interface{}); ok {
+			m = asMap[section]
+		} else {
+			return "", fmt.Errorf("invalid path %s in rollout", path)
+		}
+	}
+
+	if m == nil {
+		return "", fmt.Errorf("invalid path %s in rollout", path)
+	}
+
+	var isArray, isMap bool
+	_, isArray = m.([]interface{})
+	_, isMap = m.(map[string]interface{})
+	if isArray || isMap {
```

Comment:
what is this check is for? can you provide an example?

Reply:
If the user puts in something like `status.canary.weights.canary` - the resulting value would be a map, and not a primitive value (string/number/bool).
I guess we could support that as well, and just return it as a JSON string, but is it a reasonable usecase? It seems like it might be a future ticket.